File: workbook/projects/ch05/ll1/ll1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *input;
int pos = 0;

char lookahead() {
    return input[pos];
}

void eat(char expected) {
    if (lookahead() == expected) {
        pos++;
    } else {
        printf("Error: Expected %c but found %c\n", expected, lookahead());
        exit(1);
    }
}

void E();
void E_prime();
void T();
void T_prime();

void E() {
    T();
    E_prime();
}

void E_prime() {
    if (lookahead() == '+') {
        eat('+');
        T();
        E_prime();
    }
}

void T() {
    if (lookahead() == 'i') {  // assuming 'i' represents an integer
        eat('i');
        T_prime();
    } else if (lookahead() == '(') {
        eat('(');
        E();
        eat(')');
    }
}

void T_prime() {
    if (lookahead() == '*') {
        eat('*');
        T();
        T_prime();
    }
}

int main() {
    input = "i+i*i";  // example input
    E();
    if (lookahead() == '\0') {
        printf("Input parsed successfully!\n");
    } else {
        printf("Error: Unexpected input at end\n");
    }
    return 0;
}

File: workbook/projects/ch05/ll1/ll1.py
class LL1Parser:
    def __init__(self, input):
        self.input = input
        self.pos = 0

    def lookahead(self):
        return self.input[self.pos] if self.pos < len(self.input) else None

    def eat(self, expected):
        if self.lookahead() == expected:
            self.pos += 1
        else:
            raise Exception(f"Error: Expected {expected} but found {self.lookahead()}")

    def E(self):
        self.T()
        self.E_prime()

    def E_prime(self):
        if self.lookahead() == '+':
            self.eat('+')
            self.T()
            self.E_prime()

    def T(self):
        if self.lookahead() == 'i':  # assuming 'i' represents an integer
            self.eat('i')
            self.T_prime()
        elif self.lookahead() == '(':
            self.eat('(')
            self.E()
            self.eat(')')

    def T_prime(self):
        if self.lookahead() == '*':
            self.eat('*')
            self.T()
            self.T_prime()

    def parse(self):
        self.E()
        if self.lookahead() is None:
            print("Input parsed successfully!")
        else:
            print("Error: Unexpected input at end")

# Example usage
input_string = "i+i*i"  # example input
parser = LL1Parser(input_string)
parser.parse()
